### PR TITLE
[mle] API to provide MLE Parent Response information

### PR DESCRIPTION
**Pull request by bisdash**
This Feature allows OT to share some of the parent response stats with application when a callback for the same is installed.

### CODE REVIEW

File: include/openthread/thread.h [hunk 1]
```diff
@@ -153,6 +153,23 @@ typedef struct otIpCounters
...
 } otIpCounters;
 
+/**
+ * This structure represents the parent response stats data which a client can request on every successful parent
+ * response frame handling.
+ *
+ */
+typedef struct otThreadParentResponseStats
+{
+    otExtAddress mExtAddr;
```
[the remark this answers is not in the document]
Comment (bisdash) [reply include/openthread/thread.h]: sure

File: src/core/api/thread_api.cpp [hunk 2]
```diff
@@ -524,3 +524,20 @@ const otIpCounters *otThreadGetIp6Counters(otInstance *aInstance)
...
+#if OPENTHREAD_FTD || OPENTHREAD_MTD
+    Instance &instance = *static_cast<Instance *>(aInstance);
+
+    return instance.GetThreadNetif().GetMle().RegisterParentResponseStatsCallback(aCallback, aContext);
+#else
+    (void)aInstance;
+    (void)aCallback;
+    (void)aContext;
+
+    return OT_ERROR_NOT_IMPLEMENTED;
```
[the remark this answers is not in the document]
Comment (bisdash) [reply src/core/api/thread_api.cpp]: okay

**Comment by bisdash** [comment]
@jwhui could you please take a look at it. Abtin already reviewed it.
Biswajit

**Comment by bisdash** [comment]
+ @smides-nest

File: include/openthread/thread.h [hunk 3]
```diff
@@ -717,6 +734,30 @@ const otMleCounters *otThreadGetMleCounters(otInstance *aInstance);
...
+typedef void(OTCALL *otThreadParentResponseCallback)(otThreadParentResponseStats *aStats, void *aContext);
+
+/**
+ * This function registers a callback to receive certain stats during a parent response frame handling
+ *
+ * @param[in]  aInstance  A pointer to an OpenThread instance.
+ * @param[in]  aCallback  A pointer to a function that is called with certain parent response frame stats.
+ * @param[in]  aContext   A pointer to callback client-specific context.
+ *
+ * @retval none
```
Comment (abtink) [inline include/openthread/thread.h]: nit: document `@retval` errors?

File: src/core/thread/mle.hpp [hunk 4]
```diff
@@ -1057,6 +1057,16 @@ class Mle : public InstanceLocator
...
 
+    /**
+     * This function registers the client callback which is called upon successfully handling Parent response frame to
+     * deliver certain stats about the Parents of a node.
+     *
+     * @param[in]  aCallback A pointer to a function that is called to deliver certain stats about Parents of a node.
+     * @param[in]  aContext  A pointer to application-specific context.
+     *
+     */
+    otError RegisterParentResponseStatsCallback(otThreadParentResponseCallback aCallback, void *aContext);
```
Comment (abtink) [inline src/core/thread/mle.hpp]: Return `void`?

File: include/openthread/thread.h (hunk 3 above)
Comment (bisdash) [reply include/openthread/thread.h]: okay

File: src/core/thread/mle.hpp (hunk 4 above)
Comment (bisdash) [reply src/core/thread/mle.hpp]: okay

File: include/openthread/thread.h [hunk 5]
```diff
@@ -182,6 +182,23 @@ typedef struct otMleCounters
     uint16_t mParentChanges;
 } otMleCounters;
 
+/**
+ * This structure represents the parent response stats data which a client can request on every successful parent
+ * response frame handling.
```
Comment (smides-nest) [inline include/openthread/thread.h]: s/handling/reception/

File: include/openthread/thread.h [hunk 6]
```diff
@@ -182,6 +182,23 @@ typedef struct otMleCounters
...
+typedef struct otThreadParentResponseStats
+{
+    otExtAddress mExtAddr;      ///< IEEE 802.15.4 Extended Address of the Parent
+    uint16_t     mRloc16;       ///< Short address of the Parent
+    int8_t       mRssi;         ///< Rssi of the Parent
+    int8_t       mPriority;     ///< Parent priority
+    uint8_t      mLinkQuality3; ///< Parent Link Quality 3
+    uint8_t      mLinkQuality2; ///< Parent Link Quality 2
+    uint8_t      mLinkQuality1; ///< Parent Link Quality 1
+    uint8_t      mIsAttached;   ///< Is the node receiving parent response attached
```
Comment (smides-nest) [inline include/openthread/thread.h]: Why not bool?

File: include/openthread/thread.h [hunk 7]
```diff
@@ -182,6 +182,23 @@ typedef struct otMleCounters
     uint16_t mParentChanges;
 } otMleCounters;
 
+/**
+ * This structure represents the parent response stats data which a client can request on every successful parent
+ * response frame handling.
+ *
+ */
+typedef struct otThreadParentResponseStats
```
Comment (smides-nest) [inline include/openthread/thread.h]: Not really "stats", but the contents of the Parent Response. Maybe otThreadParentResponse{Info,Data}?

Comment (bisdash) [reply include/openthread/thread.h]: updated

File: include/openthread/thread.h (hunk 6 above)
Comment (bisdash) [reply include/openthread/thread.h]: updated

File: include/openthread/thread.h (hunk 5 above)
Comment (bisdash) [reply include/openthread/thread.h]: updated

File: src/core/thread/mle.cpp [hunk 8]
```diff
@@ -3154,6 +3156,23 @@ otError Mle::HandleParentResponse(const Message &aMessage, const Ip6::MessageInf
     }
 #endif
 
+    // Share stats with application, if requested.
+    if (mParentResponseCb)
```
Comment (jwhui) [inline src/core/thread/mle.cpp]: Should this be moved above the Partition ID checks above (e.g. above L3131)? Or is this intentionally placed after the Partition ID checks?

Comment (jwhui) [reply src/core/thread/mle.cpp]: Shouldn't this code block go above L3131 above?

Comment (bisdash) [reply src/core/thread/mle.cpp]: Yes. i m looking at my local changes and it does show that the new code is above partition-id check.  When i do a git push it says everything is up-to-date. Not sure whats wrong.Trying to figure out

Comment (bisdash) [reply src/core/thread/mle.cpp]: Was my fault.. I never commuted my change and was pushing a old branch. Now its done.

**Pull request merged**